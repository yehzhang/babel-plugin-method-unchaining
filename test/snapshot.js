import * as babel from 'babel-core';
import chai, {expect} from 'chai';
import chaiJestSnapshot from 'chai-jest-snapshot';
import {readdirSync, readFileSync} from 'fs';
import {before, beforeEach, describe, it} from 'mocha';
import rewire from 'rewire';
import {spy} from 'sinon';
import plugin from '../build/index';

chai.use(chaiJestSnapshot);

const FIXTURES_PATH = 'test/fixtures';

before(() => chaiJestSnapshot.resetSnapshotRegistry());

beforeEach(function () {
  return chaiJestSnapshot.configureUsingMochaContext(this);
});

describe('Match snapshots of generated code for', () => {
  onFixtures((fixtureName, sourceCode) => {
    it(fixtureName, () => {
      let output = babel.transform(sourceCode, {
        plugins: [plugin],
      });
      expect(output.code).to.matchSnapshot();
    });
  });
});

describe('Match snapshots of traversal for', () => {
  function spyPluginModule(pluginModule) {
    let calls = [];
    let topLevelVisitor = pluginModule.__get__('topLevelVisitor');
    topLevelVisitor.MemberExpression =
        spyMemberExpression(topLevelVisitor.MemberExpression, calls, 'Top Level');
    let unchainingVisitor = pluginModule.__get__('unchainingVisitor');
    unchainingVisitor.MemberExpression.exit =
        spyMemberExpression(unchainingVisitor.MemberExpression.exit, calls, 'Second Level');
    return calls;
  }

  function spyMemberExpression(memberExpression, calls, visitorName) {
    return function _memberExpression(path) {
      let property = path.node.computed ? path.node.property.value : path.node.property.name;
      let spyPathSkip = spy(path, 'skip');
      let result = memberExpression.call(this, path);
      calls.push({
        visitorName,
        property,
        unchained: spyPathSkip.called,
      });
      return result;
    };
  }

  onFixtures((fixtureName, sourceCode) => {
    let pluginModule = rewire('../build/index');

    it(fixtureName, () => {
      let calls = spyPluginModule(pluginModule);
      babel.transform(sourceCode, {
        plugins: [pluginModule.default],
      });

      let traversalInfo = calls.map(call => `visitor: ${call.visitorName}${call.property === null
          ? ''
          : `, property: ${call.property}`}, ${call.unchained ? 'transformed' : 'unchanged'}`);
      expect(traversalInfo).to.matchSnapshot();
    });
  });
});

function onFixtures(f) {
  readdirSync(FIXTURES_PATH)
      .filter(fixtureFile => fixtureFile !== '.DS_Store')
      .forEach(fixtureFile => {
        let fixtureName = fixtureFile.replace(/-/g, ' ');
        let fixturePath = [FIXTURES_PATH, fixtureFile].join('/');
        let sourceCode = readFileSync(fixturePath);
        f(fixtureName, sourceCode);
      });
}
