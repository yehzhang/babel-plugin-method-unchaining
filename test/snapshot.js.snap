// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Match snapshots of generated code for interleaved.js 1`] = `
"var _a;

result = (_a = (_a = Class(), _a.newBuilder).setFirst(), _a.build);
_a = undefined;"
`;

exports[`Match snapshots of generated code for multiple statements for loop body.js 1`] = `
"for (var i = 0; i < 1; i++) {
    var _a, _a2;

    result = (_a = Class(), _a.build)();
    _a = undefined;
    result = (_a2 = (_a2 = Class(), _a2.newBuilder)(), _a2.build)();
    _a2 = undefined;
}"
`;

exports[`Match snapshots of generated code for nested first level argument.js 1`] = `
"var _a;

result = (_a = (_a = (_a = Class(), _a.newBuilder)(), _a.setFirst)((_a = getDefaultInstance(), _a.clear)), _a.build);
_a = undefined;"
`;

exports[`Match snapshots of generated code for nested multiple level arguments.js 1`] = `
"var _a;

result = (_a = (_a = (_a = Class((_a = Clazz(), _a.toBuilder)), _a.newBuilder)((_a = Cls(), _a.mergeFrom)), _a.setFirst)((_a = getDefaultInstance(), _a.clear)), _a.build);
_a = undefined;"
`;

exports[`Match snapshots of generated code for nested outer argument.js 1`] = `
"var _a, _a2, _a3;

result = (_a = (_a = (_a = Class(), _a.newBuilder)(), _a.setFirst)(), _a.build)((_a2 = getDefaultInstance(), _a2.clear)((_a3 = newMessage(), _a3.toBuilder)));
_a3 = undefined;
_a2 = undefined;
_a = undefined;"
`;

exports[`Match snapshots of generated code for nested second level argument.js 1`] = `
"var _a;

result = (_a = (_a = (_a = Class(), _a.newBuilder)((_a = getDefaultInstance(), _a.clear)), _a.setFirst)(), _a.build);
_a = undefined;"
`;

exports[`Match snapshots of generated code for one level as argument.js 1`] = `
"var _a;

result = Class((_a = newBuilder(), _a.build));
_a = undefined;"
`;

exports[`Match snapshots of generated code for one level binary expression.js 1`] = `
"var _a, _a2;

result = (_a = newBuilder(), _a.build) + (_a2 = getDefaultInstance(), _a2.clear);
_a2 = undefined;
_a = undefined;"
`;

exports[`Match snapshots of generated code for one level computed.js 1`] = `
"var _a;

result = (_a = Class(), _a['newBuilder']);
_a = undefined;"
`;

exports[`Match snapshots of generated code for one level following identifier.js 1`] = `
"var _a;

result = (_a = Class.newBuilder(), _a.build);
_a = undefined;"
`;

exports[`Match snapshots of generated code for one level for loop body no braces.js 1`] = `
"for (var i = 0; i < 10; i++) {
  var _a;

  result = (_a = Class(), _a.b);
  _a = undefined;
}"
`;

exports[`Match snapshots of generated code for one level for loop variable declaration.js 1`] = `
"for (var i = (_a = Class(), _a.b); i < 10; i++) {
  var _a;
}
_a = undefined;"
`;

exports[`Match snapshots of generated code for one level ternary.js 1`] = `
"var _a, _a2;

result = result ? (_a = Class(), _a.build) : (_a2 = Class(), _a2.clear);
_a2 = undefined;
_a = undefined;"
`;

exports[`Match snapshots of generated code for one level with argument.js 1`] = `
"var _a;

result = (_a = Class(getDefaultInstance()), _a.build);
_a = undefined;"
`;

exports[`Match snapshots of generated code for one level.js 1`] = `
"var _a;

result = (_a = Class(), _a.build);
_a = undefined;"
`;

exports[`Match snapshots of generated code for three level.js 1`] = `
"var _a;

result = (_a = (_a = (_a = Class(), _a.newBuilder)(), _a.setFirst)(), _a.build);
_a = undefined;"
`;

exports[`Match snapshots of generated code for two level computed.js 1`] = `
"var _a;

result = (_a = (_a = Class(), _a['newBuilder'])(), _a['build']);
_a = undefined;"
`;

exports[`Match snapshots of generated code for two level followed by function call.js 1`] = `
"var _a;

result = (_a = (_a = newBuilder(), _a.setFirst)(), _a.build)();
_a = undefined;"
`;

exports[`Match snapshots of generated code for two level following identifier.js 1`] = `
"var _a;

result = (_a = (_a = Class.newBuilder(), _a.setFirst)(), _a.build);
_a = undefined;"
`;

exports[`Match snapshots of generated code for two level.js 1`] = `
"var _a;

result = (_a = (_a = newBuilder(), _a.setFirst)(), _a.build);
_a = undefined;"
`;

exports[`Match snapshots of generated code for unchanged.js 1`] = `
"result = Class.b;
result = Class();"
`;

exports[`Match snapshots of traversal for interleaved.js 1`] = `
Array [
  "visitor: Second Level, property: newBuilder, transformed",
  "visitor: Second Level, property: setFirst, unchanged",
  "visitor: Top Level, property: build, transformed",
]
`;

exports[`Match snapshots of traversal for multiple statements for loop body.js 1`] = `
Array [
  "visitor: Top Level, property: build, transformed",
  "visitor: Second Level, property: newBuilder, transformed",
  "visitor: Top Level, property: build, transformed",
]
`;

exports[`Match snapshots of traversal for nested first level argument.js 1`] = `
Array [
  "visitor: Second Level, property: newBuilder, transformed",
  "visitor: Second Level, property: setFirst, transformed",
  "visitor: Second Level, property: clear, transformed",
  "visitor: Top Level, property: build, transformed",
]
`;

exports[`Match snapshots of traversal for nested multiple level arguments.js 1`] = `
Array [
  "visitor: Second Level, property: toBuilder, transformed",
  "visitor: Second Level, property: newBuilder, transformed",
  "visitor: Second Level, property: mergeFrom, transformed",
  "visitor: Second Level, property: setFirst, transformed",
  "visitor: Second Level, property: clear, transformed",
  "visitor: Top Level, property: build, transformed",
]
`;

exports[`Match snapshots of traversal for nested outer argument.js 1`] = `
Array [
  "visitor: Second Level, property: newBuilder, transformed",
  "visitor: Second Level, property: setFirst, transformed",
  "visitor: Top Level, property: build, transformed",
  "visitor: Top Level, property: clear, transformed",
  "visitor: Top Level, property: toBuilder, transformed",
]
`;

exports[`Match snapshots of traversal for nested second level argument.js 1`] = `
Array [
  "visitor: Second Level, property: newBuilder, transformed",
  "visitor: Second Level, property: clear, transformed",
  "visitor: Second Level, property: setFirst, transformed",
  "visitor: Top Level, property: build, transformed",
]
`;

exports[`Match snapshots of traversal for one level as argument.js 1`] = `
Array [
  "visitor: Top Level, property: build, transformed",
]
`;

exports[`Match snapshots of traversal for one level binary expression.js 1`] = `
Array [
  "visitor: Top Level, property: build, transformed",
  "visitor: Top Level, property: clear, transformed",
]
`;

exports[`Match snapshots of traversal for one level computed.js 1`] = `
Array [
  "visitor: Top Level, property: newBuilder, transformed",
]
`;

exports[`Match snapshots of traversal for one level following identifier.js 1`] = `
Array [
  "visitor: Second Level, property: newBuilder, unchanged",
  "visitor: Top Level, property: build, transformed",
]
`;

exports[`Match snapshots of traversal for one level for loop body no braces.js 1`] = `
Array [
  "visitor: Top Level, property: b, transformed",
]
`;

exports[`Match snapshots of traversal for one level for loop variable declaration.js 1`] = `
Array [
  "visitor: Top Level, property: b, transformed",
]
`;

exports[`Match snapshots of traversal for one level ternary.js 1`] = `
Array [
  "visitor: Top Level, property: build, transformed",
  "visitor: Top Level, property: clear, transformed",
]
`;

exports[`Match snapshots of traversal for one level with argument.js 1`] = `
Array [
  "visitor: Top Level, property: build, transformed",
]
`;

exports[`Match snapshots of traversal for one level.js 1`] = `
Array [
  "visitor: Top Level, property: build, transformed",
]
`;

exports[`Match snapshots of traversal for three level.js 1`] = `
Array [
  "visitor: Second Level, property: newBuilder, transformed",
  "visitor: Second Level, property: setFirst, transformed",
  "visitor: Top Level, property: build, transformed",
]
`;

exports[`Match snapshots of traversal for two level computed.js 1`] = `
Array [
  "visitor: Second Level, property: newBuilder, transformed",
  "visitor: Top Level, property: build, transformed",
]
`;

exports[`Match snapshots of traversal for two level followed by function call.js 1`] = `
Array [
  "visitor: Second Level, property: setFirst, transformed",
  "visitor: Top Level, property: build, transformed",
]
`;

exports[`Match snapshots of traversal for two level following identifier.js 1`] = `
Array [
  "visitor: Second Level, property: newBuilder, unchanged",
  "visitor: Second Level, property: setFirst, transformed",
  "visitor: Top Level, property: build, transformed",
]
`;

exports[`Match snapshots of traversal for two level.js 1`] = `
Array [
  "visitor: Second Level, property: setFirst, transformed",
  "visitor: Top Level, property: build, transformed",
]
`;

exports[`Match snapshots of traversal for unchanged.js 1`] = `
Array [
  "visitor: Top Level, property: b, unchanged",
]
`;
